### PR TITLE
Vhbb heppy74 x genjet clustering

### LINKED ISSUE #145: re-check GenJet clustering for GenHFHadronMatcher
making an issue so I know to solve GenJet stuff from #67 which is closed.
 
from Andrea:
those are the open points in my view:

we should not remake the gen jets as they should be the same but I understand that for the hadron matching this could be actually needed? can you confirm?
in output we should have a single gen jet having all needed features, namely:
the standard flavour coming from PAT
the hadron matched flavour
the genNu correction
if hadron matcher needs to remake its own jets, then we have to back match those to the original collection and just set a had-flavour attribute

With spring15, the genjets still need to be re-clustered. As far as I know, the original authors (Johannes Hauk, @bartonaz) are taking a look since June.
 
This is the error we get halfway through the processing.
~~~
Begin processing the 171st record. Run 1, Event 2982953, LumiSection 29870 at 28-Aug-2015 11:20:24.060 CEST
----- Begin Fatal Exception 28-Aug-2015 11:20:24 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing run: 1 lumi: 29870 event: 2982953
   [1] Running path 'endpath'
   [2] Calling event method for module PoolOutputModule/'OUT'
   [3] Calling produce method for unscheduled module GenHFHadronMatcher/'matchGenBHadron'
   [4] Calling produce method for unscheduled module JetFlavourClustering/'genJetFlavourPlusLeptonInfos'
Exception Message:
InvalidID get by product ID: invalid ProductID supplied
   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
~~~

### PR DESCRIPTION
**Pull request by jpata**
Fixes #145.
- cherry-picked isAvailable fix to JetHadronMatcher
- switched to slimmedGenJets for GenHFHadronMatcher
- re-enable gen-jets in vhbb.py

Checked that matched B/C output is properly filled on genjets.

### CODE REVIEW

**Comment by jpata** [comment]
I've realized there is a problem with this commit. `event.genJets` are actually already read and filtered by the heppy JetAnalyzer. Here I read them again because the GenHFHadronMatcher produces output which is indexed to the original genjet vector.

In the end the output is a bit weird, with always nGenJets==15 which is not correct.
~~~
(Long64_t) 25
root [2] tree->Scan("nGenJet:GenJet_pt:GenJet_numBHadrons")
***********************************************************
*    Row   * Instance *   nGenJet * GenJet_pt * GenJet_nu *
***********************************************************
*        0 *        0 *        15 * 0.0247375 *         1 *
*        0 *        1 *        15 * 0.7395740 *         0 *
*        0 *        2 *        15 * -1.153265 *         0 *
*        0 *        3 *        15 * 0.3924851 *         1 *
*        0 *        4 *        15 * 1.6125000 *         0 *
*        0 *        5 *        15 * -1.254936 *         0 *
*        0 *        6 *        15 *       -99 *       -99 *
*        0 *        7 *        15 *       -99 *       -99 *
*        0 *        8 *        15 *       -99 *       -99 *
~~~

I will fix by making a PR against Heppy where the jet index inside the original vector is saved in the genJet.

**Comment by jpata** [comment]
The genJets were declared in duplicate in both `vhbb.py` and `vhbb-combined.py` which confused the auto-ntuplizer. Now it's fixed, we also read the genjets directly from the event and use their corresponding index. Currently I put the heppy-side indexing path here as well as cherry-picking it onto @cbernet's 747 branch (https://github.com/cbernet/cmssw/pull/51).
@arizzi Please let me know if that's OK or not.

**Comment by arizzi** [comment]
@jpata can you rebase please? it is no more mergeable

File: PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py [hunk 1]
```diff
@@ -113,6 +113,8 @@ def process(self, event):
 
         if self.cfg_comp.isMC:
             self.genJets = [ x for x in self.handles['genJet'].product() ]
+            for igj, gj in enumerate(self.genJets):
+                gj.index = igj
```
Comment (arizzi) [inline PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py]: do we really want to put it in central heppy?

Comment (arizzi) [inline PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py]: explained via chat. it's ok I'll take care of PR on central heppy

**Comment by arizzi** [comment]
Heppy side already in cbernet/cmssw#51

**Pull request merged**